server/src/main/java/org/apache/iotdb/db/utils/QueryDataSetUtils.java

```diff
@@ -66,14 +67,57 @@ public static TSQueryDataSet convertQueryDataSetByFetchSize(
     for (int i = 0; i < fetchSize; i++) {
       if (queryDataSet.hasNext()) {
         RowRecord rowRecord = queryDataSet.next();
-        // filter rows whose columns are null according to the rule
-        if ((queryDataSet.isWithoutAllNull() && rowRecord.isAllNull())
-            || (queryDataSet.isWithoutAnyNull() && rowRecord.hasNullField())) {
-          // if the current RowRecord doesn't satisfy, we should also decrease AlreadyReturnedRowNum
-          queryDataSet.decreaseAlreadyReturnedRowNum();
-          i--;
-          continue;
+        Set<Integer> withoutNullColumnsIndex = queryDataSet.getWithoutNullColumnsIndex();
+        if (withoutNullColumnsIndex.isEmpty()) {
+          // filter rows whose columns are null according to the rule
+          if ((queryDataSet.isWithoutAllNull() && rowRecord.isAllNull())
+              || (queryDataSet.isWithoutAnyNull() && rowRecord.hasNullField())) {
+            // if the current RowRecord doesn't satisfy, we should also decrease
+            // AlreadyReturnedRowNum
+            queryDataSet.decreaseAlreadyReturnedRowNum();
+            i--;
+            continue;
+          }
+        } else {
+          boolean anyNullFlag = false, allNullFlag = true;
+          int index = 0;
+          for (Field field : rowRecord.getFields()) {
+            if (!withoutNullColumnsIndex.contains(index)) {
```

Comment:
`withoutNullColumnsIndex` should not be very large, but the number of rowRecord's fields may be large. Exchange the loop sequence

Reply:
fix